build_barback/lib/src/transformer/crawl_imports.dart

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+import 'dart:async';
+
+import 'package:analyzer/analyzer.dart' show parseDirectives;
+import 'package:analyzer/src/generated/engine.dart' show TimestampedData;
+import 'package:analyzer/src/generated/source.dart';
+import 'package:build/build.dart';
+import 'package:path/path.dart' as p;
+
+import 'package:analyzer/dart/ast/ast.dart';
+
+/// Crawl Dart files outside of the SDK transitively imported from any library
+/// in [entryPoints] and cache it as a [Source].
+Future<Map<Uri, Source>> crawlImports(
+    Iterable<AssetId> entryPoints, Future<String> read(AssetId id)) async {
+  var roots = entryPoints.where((id) => id.path.endsWith('.dart'));
+  var assets = await _crawlAsync(roots, read, _assetImports);
+  var results = <Uri, Source>{};
+  for (var asset in assets.keys) {
+    results[asset.uri] = _toSource(asset, assets[asset]);
+  }
+  return results;
+}
+
+Iterable<AssetId> _assetImports(AssetId id, String source) {
+  var unit = parseDirectives(source);
+  return unit.directives
+      .where((d) => d is UriBasedDirective)
+      .map((d) => d as UriBasedDirective)
+      .map((d) => d.uri.stringValue)
+      .where((uri) => !uri.startsWith('dart:'))
+      .map((uri) => new AssetId.resolve(uri, from: id));
+}
+
+Source _toSource(AssetId id, String content) => new AssetSource(id, content);
+
+/// Crawl a graph where node values and edges are resolved asynchronously.
+///
+/// Nodes are keyed by type [K] and have values of type [V]. [resolve] finds the
+/// value for a node. [edges] finds the edges leaving a node.
```

Comment:
Maybe add something like `Returns all nodes found by crawling [roots]` (it wasn't super clear to me what the return value represented right away).